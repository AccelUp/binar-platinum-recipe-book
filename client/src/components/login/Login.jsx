import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import classes from "./Login.module.css";

const Login = () => {
  const navigate = useNavigate();
  const [notification, setNotification] = useState(null);
  const [errors, setErrors] = useState({});
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });

  const navigateToSignUp = () => {
    navigate("/signup");
  };

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const fetchLogin = async (e) => {
    e.preventDefault();
    const newErrors = {};

    if (formData.username.trim() === "") {
      newErrors.username = "Username is required";
    }

    if (formData.password.trim() === "") {
      newErrors.password = "Password is required";
    }

    setErrors(newErrors);

    if (Object.keys(newErrors).length === 0) {
      try {
        const response = await fetch(
          "http://localhost:8000/api/v1/recipebook/user/login",
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(formData),
          }
        );

        if (response.ok) {
          navigate("/dashboard");
        } else {
          setNotification(
            "Failed to login. Please check your username and password."
          );
        }
      } catch (error) {
        console.error("Failed to fetch user data.", error);
      }
    }
  };

  return (
    <div className={`${classes.Login} relative pt-10`}>
      <div className="h-screen flex justify-center items-center">
        <div className="bg-black mx-4 p-8 rounded shadow-md w-full md:w-1/2 lg:w-1/3">
          <h1 className="text-primary font-bold mb-8 text-center">Login</h1>
          {notification && (
            <div className="text-red-500 mb-4">{notification}</div>
          )}
          <form onSubmit={fetchLogin}>
            <div className="mb-4">
              <label
                className="block font-semibold text-primary mb-2"
                htmlFor="username"
              >
                Username
              </label>
              <input
                className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                  errors.username ? "border-red-500" : ""
                }`}
                id="username"
                name="username"
                value={formData.username}
                type="text"
                onChange={handleInputChange}
                placeholder="Enter your username"
              />
              {errors.username && (
                <p className="text-red-500 text-xs">{errors.username}</p>
              )}
            </div>
            <div className="mb-4">
              <label
                className="block font-semibold text-primary mb-2"
                htmlFor="password"
              >
                Password
              </label>
              <input
                className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                  errors.password ? "border-red-500" : ""
                }`}
                id="password"
                name="password"
                type="password"
                value={formData.password}
                onChange={handleInputChange}
                placeholder="Enter your password"
              />
              {errors.password && (
                <p className="text-red-500 text-xs">{errors.password}</p>
              )}
              <h2 className="mb-2 text-sm pt-2">
                Don't have an account?
                <span
                  onClick={navigateToSignUp}
                  className="text-primary cursor-pointer"
                >
                  Sign up
                </span>
              </h2>
            </div>
            <div className="mb-6">
              <button
                className="bg-primary hover:bg-black hover:text-primary text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

export default Login;
