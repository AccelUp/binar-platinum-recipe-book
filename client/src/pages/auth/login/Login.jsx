import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { login } from "../../../actions/authAction";
import classes from "./Login.module.css";

const Login = () => {
  const navigate = useNavigate();
  const [errorText, setErrorText] = useState(""); // State untuk pesan kesalahan
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });

  const dispatch = useDispatch();

  const navigateToSignUp = () => {
    navigate("/signup");
  };

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const fetchLogin = async (e) => {
    e.preventDefault();

    if (formData.username === "" || formData.password === "") {
      setErrorText("Username and password are required"); // Set pesan kesalahan
    } else {
      try {
        const response = await dispatch(login(formData, navigate));
        if (response.error) {
          setErrorText("User not found. Please check your credentials.");
        }
      } catch (error) {
        console.error("Failed to fetch user data", error);
        setErrorText("An error occurred while logging in.");
      }
    }
  };

  return (
    <div className={`${classes.Login} relative pt-10`}>
      <div className="h-screen flex justify-center items-center">
        <div className="bg-black mx-4 p-8 rounded shadow-md w-full md:w-1/2 lg:w-1/3">
          <h1 className="text-primary font-bold mb-8 text-center">Login</h1>
          {errorText && ( // Tampilkan pesan kesalahan jika ada
            <div className="text-red-500 mb-4 p-2 bg-red-100 border border-red-400 rounded">
              {errorText}
            </div>
          )}
          <form onSubmit={fetchLogin}>
            <div className="mb-4">
              <label
                className="block font-semibold text-primary mb-2"
                htmlFor="username"
              >
                Username
              </label>
              <input
                className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                  errorText ? "border-red-500" : ""
                }`}
                id="username"
                name="username"
                value={formData.username}
                type="text"
                onChange={handleInputChange}
                placeholder="Enter your username"
              />
            </div>
            <div className="mb-4">
              <label
                className="block font-semibold text-primary mb-2"
                htmlFor="password"
              >
                Password
              </label>
              <input
                className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                  errorText ? "border-red-500" : ""
                }`}
                id="password"
                name="password"
                type="password"
                value={formData.password}
                onChange={handleInputChange}
                placeholder="Enter your password"
              />
            </div>
            <div className="mb-6">
              <button
                className="bg-primary hover:bg-black hover:text-primary text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
          <h2 className="mb-2 text-sm pt-2">
            Don't have an account?{" "}
            <span
              onClick={navigateToSignUp}
              className="text-primary cursor-pointer"
            >
              Sign up
            </span>
          </h2>
        </div>
      </div>
    </div>
  );
};

export default Login;
