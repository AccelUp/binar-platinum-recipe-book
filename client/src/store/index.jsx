import { createStore, applyMiddleware } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";
import rootReducer from "../reducers";
import { composeWithDevTools } from "redux-devtools-extension";

// Konfigurasi Redux Persist
const persistConfig = {
  key: "root", // Key untuk penyimpanan
  storage, // Jenis penyimpanan (local storage)
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Buat store dengan middleware dan reducer yang telah di-persist
const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(thunk))
);
const persistor = persistStore(store);

export { store, persistor };
