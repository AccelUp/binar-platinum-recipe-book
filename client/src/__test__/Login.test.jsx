import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../store";
import Login from "../pages/auth/login/Login";
import renderer from "react-test-renderer";

describe("Login Component", () => {
  it("should render the Login component correctly", () => {
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
