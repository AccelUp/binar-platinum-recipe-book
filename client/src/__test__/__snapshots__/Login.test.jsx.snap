// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Component > should render the Login component correctly 1`] = `
<div
  className="_Login_17c26c relative pt-10"
>
  <div
    className="h-screen flex justify-center items-center"
  >
    <div
      className="bg-black mx-4 p-8 rounded shadow-md w-full md:w-1/2 lg:w-1/3"
    >
      <h1
        className="text-primary font-bold mb-8 text-center"
      >
        Login
      </h1>
      <form
        onSubmit={[Function]}
      >
        <div
          className="mb-4"
        >
          <label
            className="block font-semibold text-primary mb-2"
            htmlFor="username"
          >
            Username
          </label>
          <input
            className="border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Enter your username"
            type="text"
            value=""
          />
        </div>
        <div
          className="mb-4"
        >
          <label
            className="block font-semibold text-primary mb-2"
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
            id="password"
            name="password"
            onChange={[Function]}
            placeholder="Enter your password"
            type="password"
            value=""
          />
        </div>
        <div
          className="mb-6"
        >
          <button
            className="bg-primary hover:bg-black hover:text-primary text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
      <h2
        className="mb-2 text-sm pt-2"
      >
        Don't have an account?
         
        <span
          className="text-primary cursor-pointer"
          onClick={[Function]}
        >
          Sign up
        </span>
      </h2>
    </div>
  </div>
</div>
`;
